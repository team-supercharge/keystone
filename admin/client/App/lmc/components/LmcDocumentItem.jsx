import React, { Component } from 'react'
import PropTypes from 'prop-types'
import moment from 'moment'
import { isBrowser, isTablet } from 'react-device-detect'
import { Link } from 'react-router'
import { Button, GlyphButton } from '../../elemental'
import ConfirmationDialog from '../../shared/ConfirmationDialog'

export default class LmcDocumentItem extends Component {
    state = {
        deleteDialogOpen: false
    }
    
    toggleDeleteDialog = () => {
        this.setState(prevState => ({
            deleteDialogOpen: !prevState.deleteDialogOpen
        }))
    }

    handleDeleteConfirm = (id) => {
        this.toggleDeleteDialog()
        this.props.onDelete(id)
    }

    render () {
        const { data, listId } = this.props
        const daysDiff = moment().diff(data.createdAt, 'days')
        const displayedTime = daysDiff <= 7 
            ? moment(data.createdAt).calendar() 
            : `Added ${daysDiff} days ago`
        const editLink = `${Keystone.adminPath}/${listId}/${data.id}`
        const chosenStyles = (isBrowser || isTablet) ? desktopStyles : mobileStyles
        return (
            <li style={chosenStyles.container}>
                <div style={chosenStyles.infoContainer}>
                    <div style={desktopStyles.imageContainer}>
                        <img 
                            src={IMAGE_URL} 
                            style={desktopStyles.image} 
                        />
                    </div>
                    <div style={desktopStyles.textContainer}>
                        <a 
                            href={data.pdf}
                            style={desktopStyles.documentName}
                            target='_blank'
                        >
                            {data.name}
                        </a>
                        <br />
                        <span style={desktopStyles.dateDiff}>
                            {displayedTime}
                        </span>
                    </div>
                </div>
                <div style={chosenStyles.buttonContainer}>
                    <Button
                        style={chosenStyles.button} 
                        color='default'
                        href={data.pdf}
                        target='_blank'
                    >
                        View
                    </Button>
                    <GlyphButton
                        component={Link}
                        glyph='pencil'
                        position='left'
                        style={chosenStyles.button}
                        to={editLink}
                    >
                        Edit
                    </GlyphButton>
                    <Button 
                        onClick={this.toggleDeleteDialog} 
                        variant="link" 
                        color="delete"
                        style={chosenStyles.button} 
                        data-button="delete"
                    >
                        Delete
                    </Button>
                </div>
                <ConfirmationDialog
                    confirmationLabel='Delete'
                    confirmationType='warning'
                    isOpen={this.state.deleteDialogOpen}
                    onCancel={this.toggleDeleteDialog}
                    onConfirmation={() => this.handleDeleteConfirm(data.id)}
                >
                    <div>
                        Are you sure you want to delete this document? 
                        If you go ahead, it can’t be undone. 
                        Once it's gone, it’s gone for good!
                    </div>
                </ConfirmationDialog>
            </li>
        )
    }
}

const IMAGE_URL = 'https://s3.eu-west-2.amazonaws.com/lmc-data-production/public/assessment.png'

const desktopStyles = {
    button: {
        position: 'relative',
        top: 5,
        marginLeft: 10,
    },
    buttonContainer: {
        display: 'flex',
        flexDirection: 'row',
        justifyContent: 'flex-end',
    },
    container: {
        display: 'flex',
        flexDirection: 'row',
        justifyContent: 'space-between',
        marginBottom: 8,
    },
    dateDiff: {
        fontSize: 12,
        opacity: '0.6',
    },
    documentName: {
        color: 'black',
        fontSize: 15,
        textDecoration: 'none',
    },
    image: {
        width: 20,
        height: 20,
        margin: '9px 0px 0px 0px'
    },
    imageContainer: {
        width: 40,
        height: 40,
        borderRadius: 40 / 2,
        textAlign: 'center',
        backgroundColor: '#8AABC8',
        float: 'left',
    },
    textContainer: {
        display: 'inline-block',
        paddingLeft: 20,
    },
}

const mobileStyles = {
    container: {
        display: 'flex',
        flexDirection: 'column'
    },
    button: {
        float: 'left',
        position: 'relative',
        top: 5,
        marginRight: 10,
    },
    buttonContainer: {
        display: 'flex',
        flexDirection: 'row',
        paddingTop: 8,
        marginBottom: 8,
        marginLeft: 59,
    },
}

LmcDocumentItem.propTypes = {
    data: PropTypes.object.isRequired,
    listId: PropTypes.string.isRequired,
    onDelete: PropTypes.func.isRequired,
}