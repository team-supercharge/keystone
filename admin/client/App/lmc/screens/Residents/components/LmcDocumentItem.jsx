import React, { Component } from 'react'
import PropTypes from 'prop-types'
import moment from 'moment'
import { Button } from '../../../../elemental'
import ConfirmationDialog from '../../../../shared/ConfirmationDialog'

export default class LmcDocumentItem extends Component {
    state = {
        deleteDialogOpen: false
    }
    
    toggleDeleteDialog = () => {
        this.setState(prevState => ({
            deleteDialogOpen: !prevState.deleteDialogOpen
        }))
    }

    render () {
        const { data, onDelete } = this.props
        const currentDate = moment()
        const diff = currentDate.diff(data.createdAt, 'days')

        return (
            <li style={styles.container}>
                <div style={styles.imageContainer}>
                    <img 
                        src={IMAGE_URL} 
                        style={styles.image} 
                    />
                </div>
                <div style={styles.textContainer}>
                    <a 
                        href={data.pdf}
                        style={styles.documentName}
                        target='_blank'
                    >
                        {data.name}
                    </a>
                    <br />
                    <span style={styles.dateDiff}>
                        Added {diff} days ago
                    </span>
                </div>
                <Button 
                    onClick={this.toggleDeleteDialog} 
                    variant="link" 
                    color="delete"
                    style={styles.deleteButton} 
                    data-button="delete"
                >
                    Delete
                </Button>
                <Button
                    style={styles.viewButton} 
                    color='default'
                    href={data.pdf}
                    target='_blank'
                >
                    View
                </Button>
                <ConfirmationDialog
                    confirmationLabel='Delete'
                    confirmationType='danger'
                    isOpen={this.state.deleteDialogOpen}
                    onCancel={this.toggleDeleteDialog}
                    onConfirmation={() => onDelete(data.id)}
                >
                    <div>
                        Are you sure you want to delete this document? If you go ahead, it can’t be undone. Once it's gone, it’s gone for good!
                    </div>
                </ConfirmationDialog>
            </li>
        )
    }
}

const IMAGE_URL = 'https://s3.eu-west-2.amazonaws.com/lmc-data-production/public/assessment.png'

const styles = {
    container: {
        height: 70,
    },
    dateDiff: {
        fontSize: 12,
        opacity: '0.6',
    },
    deleteButton: {
        float: 'right'
    },
    documentName: {
        color: 'black',
        fontSize: 15,
        textDecoration: 'none',
    },
    image: {
        width: 20,
        height: 20,
        margin: '9px 0px 0px 0px'
    },
    imageContainer: {
        width: 40,
        height: 40,
        borderRadius: 40 / 2,
        textAlign: 'center',
        backgroundColor: '#8AABC8',
        float: 'left',
        position: 'relative',
        bottom: 2,
    },
    textContainer: {
        display: 'inline-block',
        paddingLeft: 20,
    },
    viewButton: {
        float: 'right',
        padding: '0px !important',
    }
}

LmcDocumentItem.propTypes = {
    data: PropTypes.object.isRequired,
    onDelete: PropTypes.func.isRequired,
}