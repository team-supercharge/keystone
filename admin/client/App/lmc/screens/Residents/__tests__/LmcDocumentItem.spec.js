import React from 'react'
import { shallow } from 'enzyme'
import MockDate from 'mockdate'
import LmcDocumentItem from '../components/LmcDocumentItem';

MockDate.set('1/1/2019')

describe('LmcDocumentItem', () => {
    let wrapper
    let data
    let savedKeystone
    const onDelete = jest.fn()

    beforeEach(() => {
        savedKeystone = global.Keystone
        global.Keystone = { adminPath: '/admin' }

        data = {
            name: 'TestDocument',
            pdf: 'TestLink',
            id: 'TestId'
        }
        wrapper = shallow(
            <LmcDocumentItem
                data={data}
                onDelete={onDelete}
            />
        )
    })

    it('renders correctly', () => {
        expect(wrapper).toMatchSnapshot()
    })

    it('renders a list item', () => {
        expect(wrapper.find('li').length).toEqual(1)
    })

    it('displays a link with the data name to open the document in a new tab', () => {
        const link = wrapper.find('a')
        expect(link.props().href).toEqual(data.pdf)
        expect(link.props().children).toEqual(data.name)
        expect(link.props().target).toEqual('_blank')
    })

    it('displays a button to open the document in a new tab', () => {
        const button = wrapper.find('Button').at(1)
        expect(button.props().href).toEqual(data.pdf)
        expect(button.props().target).toEqual('_blank')
        expect(button.props().children).toEqual('View')
    })

    it("has a 'pencil' button to edit the document", () => {
        const editButton = wrapper.find('GlyphButton')
        expect(editButton.props().to).toEqual('/admin/documents/TestId')
        expect(editButton.props().glyph).toEqual('pencil')
    })

    it('has a confirmation dialog to delete the data', () => {
        const dialog = wrapper.find('ConfirmationDialog')
        expect(dialog.props().confirmationType).toEqual('danger')
        expect(dialog.props().confirmationLabel).toEqual('Delete')
    })

    it('executes its onDelete function when confirming the delete dialog', () => {
        const dialog = wrapper.find('ConfirmationDialog')
        dialog.props().onConfirmation()
        expect(onDelete).toBeCalledTimes(1)
    })

    afterEach(() => {
        onDelete.mockClear()
    })

    afterAll(() => {
        MockDate.clear()
        global.Keystone = savedKeystone
    })
})