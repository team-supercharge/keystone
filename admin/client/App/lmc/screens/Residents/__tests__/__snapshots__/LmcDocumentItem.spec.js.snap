// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LmcDocumentItem renders correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <li
    style={
      Object {
        "height": 70,
      }
    }
  >
    <div
      style={
        Object {
          "backgroundColor": "#8AABC8",
          "borderRadius": 20,
          "bottom": 2,
          "float": "left",
          "height": 40,
          "position": "relative",
          "textAlign": "center",
          "width": 40,
        }
      }
    >
      <img
        src="https://s3.eu-west-2.amazonaws.com/lmc-data-production/public/assessment.png"
        style={
          Object {
            "height": 20,
            "margin": "9px 0px 0px 0px",
            "width": 20,
          }
        }
      />
    </div>
    <div
      style={
        Object {
          "display": "inline-block",
          "paddingLeft": 20,
        }
      }
    >
      <a
        href="TestLink"
        style={
          Object {
            "color": "black",
            "fontSize": 15,
            "textDecoration": "none",
          }
        }
        target="_blank"
      >
        TestDocument
      </a>
      <br />
      <span
        style={
          Object {
            "fontSize": 12,
            "opacity": "0.6",
          }
        }
      >
        Added 
        0
         days ago
      </span>
    </div>
    <Button
      aphroditeStyles={Array []}
      color="delete"
      data-button="delete"
      onClick={[Function]}
      style={
        Object {
          "float": "right",
        }
      }
      variant="link"
    >
      Delete
    </Button>
    <Button
      aphroditeStyles={Array []}
      color="default"
      href="TestLink"
      style={
        Object {
          "float": "right",
          "padding": "0px !important",
        }
      }
      target="_blank"
      variant="fill"
    >
      View
    </Button>
    <ConfirmationDialog
      cancelLabel="Cancel"
      confirmationLabel="Delete"
      confirmationType="danger"
      isOpen={false}
      onCancel={[Function]}
      onConfirmation={[Function]}
    >
      <div>
        Are you sure you want to delete this document? If you go ahead, it can’t be undone. Once it's gone, it’s gone for good!
      </div>
    </ConfirmationDialog>
  </li>,
  "nodes": Array [
    <li
      style={
        Object {
          "height": 70,
        }
      }
    >
      <div
        style={
          Object {
            "backgroundColor": "#8AABC8",
            "borderRadius": 20,
            "bottom": 2,
            "float": "left",
            "height": 40,
            "position": "relative",
            "textAlign": "center",
            "width": 40,
          }
        }
      >
        <img
          src="https://s3.eu-west-2.amazonaws.com/lmc-data-production/public/assessment.png"
          style={
            Object {
              "height": 20,
              "margin": "9px 0px 0px 0px",
              "width": 20,
            }
          }
        />
      </div>
      <div
        style={
          Object {
            "display": "inline-block",
            "paddingLeft": 20,
          }
        }
      >
        <a
          href="TestLink"
          style={
            Object {
              "color": "black",
              "fontSize": 15,
              "textDecoration": "none",
            }
          }
          target="_blank"
        >
          TestDocument
        </a>
        <br />
        <span
          style={
            Object {
              "fontSize": 12,
              "opacity": "0.6",
            }
          }
        >
          Added 
          0
           days ago
        </span>
      </div>
      <Button
        aphroditeStyles={Array []}
        color="delete"
        data-button="delete"
        onClick={[Function]}
        style={
          Object {
            "float": "right",
          }
        }
        variant="link"
      >
        Delete
      </Button>
      <Button
        aphroditeStyles={Array []}
        color="default"
        href="TestLink"
        style={
          Object {
            "float": "right",
            "padding": "0px !important",
          }
        }
        target="_blank"
        variant="fill"
      >
        View
      </Button>
      <ConfirmationDialog
        cancelLabel="Cancel"
        confirmationLabel="Delete"
        confirmationType="danger"
        isOpen={false}
        onCancel={[Function]}
        onConfirmation={[Function]}
      >
        <div>
          Are you sure you want to delete this document? If you go ahead, it can’t be undone. Once it's gone, it’s gone for good!
        </div>
      </ConfirmationDialog>
    </li>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <LmcDocumentItem
        data={
          Object {
            "name": "TestDocument",
            "pdf": "TestLink",
          }
        }
        onDelete={[MockFunction]}
      />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": LmcDocumentItem {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "data": Object {
            "name": "TestDocument",
            "pdf": "TestLink",
          },
          "onDelete": [MockFunction],
        },
        "refs": Object {},
        "state": Object {
          "deleteDialogOpen": false,
        },
        "toggleDeleteDialog": [Function],
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <li
          style={
            Object {
              "height": 70,
            }
          }
        >
          <div
            style={
              Object {
                "backgroundColor": "#8AABC8",
                "borderRadius": 20,
                "bottom": 2,
                "float": "left",
                "height": 40,
                "position": "relative",
                "textAlign": "center",
                "width": 40,
              }
            }
          >
            <img
              src="https://s3.eu-west-2.amazonaws.com/lmc-data-production/public/assessment.png"
              style={
                Object {
                  "height": 20,
                  "margin": "9px 0px 0px 0px",
                  "width": 20,
                }
              }
            />
          </div>
          <div
            style={
              Object {
                "display": "inline-block",
                "paddingLeft": 20,
              }
            }
          >
            <a
              href="TestLink"
              style={
                Object {
                  "color": "black",
                  "fontSize": 15,
                  "textDecoration": "none",
                }
              }
              target="_blank"
            >
              TestDocument
            </a>
            <br />
            <span
              style={
                Object {
                  "fontSize": 12,
                  "opacity": "0.6",
                }
              }
            >
              Added 
              0
               days ago
            </span>
          </div>
          <Button
            aphroditeStyles={Array []}
            color="delete"
            data-button="delete"
            onClick={[Function]}
            style={
              Object {
                "float": "right",
              }
            }
            variant="link"
          >
            Delete
          </Button>
          <Button
            aphroditeStyles={Array []}
            color="default"
            href="TestLink"
            style={
              Object {
                "float": "right",
                "padding": "0px !important",
              }
            }
            target="_blank"
            variant="fill"
          >
            View
          </Button>
          <ConfirmationDialog
            cancelLabel="Cancel"
            confirmationLabel="Delete"
            confirmationType="danger"
            isOpen={false}
            onCancel={[Function]}
            onConfirmation={[Function]}
          >
            <div>
              Are you sure you want to delete this document? If you go ahead, it can’t be undone. Once it's gone, it’s gone for good!
            </div>
          </ConfirmationDialog>
        </li>,
        "_debugID": 2,
        "_renderedOutput": <li
          style={
            Object {
              "height": 70,
            }
          }
        >
          <div
            style={
              Object {
                "backgroundColor": "#8AABC8",
                "borderRadius": 20,
                "bottom": 2,
                "float": "left",
                "height": 40,
                "position": "relative",
                "textAlign": "center",
                "width": 40,
              }
            }
          >
            <img
              src="https://s3.eu-west-2.amazonaws.com/lmc-data-production/public/assessment.png"
              style={
                Object {
                  "height": 20,
                  "margin": "9px 0px 0px 0px",
                  "width": 20,
                }
              }
            />
          </div>
          <div
            style={
              Object {
                "display": "inline-block",
                "paddingLeft": 20,
              }
            }
          >
            <a
              href="TestLink"
              style={
                Object {
                  "color": "black",
                  "fontSize": 15,
                  "textDecoration": "none",
                }
              }
              target="_blank"
            >
              TestDocument
            </a>
            <br />
            <span
              style={
                Object {
                  "fontSize": 12,
                  "opacity": "0.6",
                }
              }
            >
              Added 
              0
               days ago
            </span>
          </div>
          <Button
            aphroditeStyles={Array []}
            color="delete"
            data-button="delete"
            onClick={[Function]}
            style={
              Object {
                "float": "right",
              }
            }
            variant="link"
          >
            Delete
          </Button>
          <Button
            aphroditeStyles={Array []}
            color="default"
            href="TestLink"
            style={
              Object {
                "float": "right",
                "padding": "0px !important",
              }
            }
            target="_blank"
            variant="fill"
          >
            View
          </Button>
          <ConfirmationDialog
            cancelLabel="Cancel"
            confirmationLabel="Delete"
            confirmationType="danger"
            isOpen={false}
            onCancel={[Function]}
            onConfirmation={[Function]}
          >
            <div>
              Are you sure you want to delete this document? If you go ahead, it can’t be undone. Once it's gone, it’s gone for good!
            </div>
          </ConfirmationDialog>
        </li>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <LmcDocumentItem
    data={
      Object {
        "name": "TestDocument",
        "pdf": "TestLink",
      }
    }
    onDelete={[MockFunction]}
  />,
}
`;
